### ISSUE TITLE
Preventing very high email traffic by many bans (ex. DDoS attack)

### ISSUE THREAD

**Issue by szepeviktor** [fail2ban/fail2ban#832]
It would be nice to my server to detect ongoing DDoS and stop sending notification emails but a last special one saying **Silencing further notifications**.

What is the way of detecting a DDoS?

Sample attack: ![DDoS](http://i.snag.gy/dB2VD.jpg)

**Comment by leeclemens** [comment]
I'm confused about the issue here. Do you want to receive an email per banned IP? If so, this has nothing to do with a DDoS, but that you have a lot of IPs being banned and wish to receive an email about each one. Rate limiting a specific action is something altogether.

**Comment by szepeviktor** [comment]
On DDoS (very frequent emails genetated by fail2ban) I'd like to receive a digest only.
Sending emails is also a load on the server.

**Comment by szepeviktor** [comment]
Note that I use fail2ban to ban on various actions (404s, url hacks, failed AJAX, php in uploads folder) not just authentication failures.

**Comment by leeclemens** [comment]
When you say DDoS, are you referring to your server running fail2ban and sending emails to yourself?

**Comment by szepeviktor** [comment]
This is not a classical DDoS but a mini-DDoS, and I have a *mini* server.

![DDoS](http://i.snag.gy/4t9zx.jpg)

This is a real HTTP attack caught by [wpf2b](https://github.com/szepeviktor/wordpress-plugin-construction/tree/master/wordpress-fail2ban).

**Comment by szepeviktor** [comment]
So an actual DDoS causes even more load on your system by sending emails.
In a DDoS scenario you need every bit of resources.

**Comment by leeclemens** [comment]
So ~100 new hosts per day got blocked? Is fail2ban blocking them at the firewall, or from the application (wordpress)? Was fail2ban causing resource exhaustion, or sendmail?

**Comment by szepeviktor** [comment]
Of course at `iptables`.
I was creating this issue to give more resources to Apache and to save me from 500 emails (see the first comment). Apache survived this attack. wpf2b runs (end exits on ban) before WordPress loads so my 1.5 GB of memory was enough.

**Comment by leeclemens** [comment]
So what is the issue, your title refers to a DDoS attack, and your previous comment insinuates it is fail2ban which is 'DDoSing' your server (you didn't directly answer my question earlier)? I'm merely trying to help make fail2ban better and in the process help you; to do so I need to better understand the issue you experience.

If you are sending 500 emails, is that what is using too many resources? Are they queued?

**Comment by szepeviktor** [comment]
The original title was [Detecting ongoing DDoS attack](https://github.com/fail2ban/fail2ban/issues/832#event-183669039)

> your previous comment insinuates it is fail2ban which is 'DDoSing' your server

No, fail2ban caused a bit more load on my server and overloaded my email account. I was always referring real HTTP DDoS attacks.

My very small VPS with spinning disks could handle 500 emails but I see those emails as superfluous because all of them were about one specific kind of attack from various IP addresses.

I'd like to stop emails during an ongoing DDoS attack because:

- No new information can be found in the emails
- Generating emails degrades the overall performance of the server
- It takes time to delete hundreds of almost-identical emails

It would be nice stop emails per jail so other jails could still trigger emails.

**Comment by szepeviktor** [comment]
Thank you for caring about my problem.
I hope others are also irritated about 500 emails.

**Comment by leeclemens** [comment]
Sorry if I had misunderstood what you were referring to as the "DDoS". I don't think a lot of apache-noscript attempts make up a DDoS, but it's not useful to split hairs regarding this.

I think a general breakdown of what this Issue is about doesn't need to consider "DDoS", as it isn't really related.

Given:
A jail is configured to send email notifications per ban
Lots of hosts are banned -> Lots of emails are generated

Feature Request:
Digest/rate limit email notifications (per jail, preferably)

Question:
What do you mean by "No new information can be found in the emails" ?

**Comment by sebres** [comment]
@leeclemens Lee, please read my comment above, about a possible FR "prevent very many emails": https://github.com/fail2ban/fail2ban/issues/832#issuecomment-60483697 
I will push (possibly next week) a possible solution to this issue.

**Comment by szepeviktor** [comment]
> What do you mean by "No new information can be found in the emails" ?

@leeclemens Those attacks were all uniform WordPress login probes to the same site. The only *new* information is the IP address.
Beacuse of the attack uniformity I knew that it was a botnet.

**Comment by leeclemens** [comment]
@sebres I would hate for this to be the default (delayed notification) - can it be set to 0m by default to mirror the currently expected behavior of the action being performed when a ban occurs?

**Comment by szepeviktor** [comment]
@leeclemens Don't you say you like your inbox with hunders of almost identical emails?

**Comment by leeclemens** [comment]
@szepeviktor I don't think I said that...do you have a link to a comment? If I configure fail2ban to send me notifications - I would expect it to send me notifications. If it did not send me notifications, I would be concerned it was not functioning as I had configured it.

My concern here is, if the default behavior is changed just because someones server keels over sending 100 emails a day - the owner of which deliberately configured fail2ban to send the emails in the first place. If something like a digest can be implemented, that is great! I just don't want the default to cater to a corner case.

**Comment by szepeviktor** [comment]
OK

**Comment by leeclemens** [comment]
Perhaps another action altogether would be better? Or a wrapper for actions that send mail? I'm just thinking out loud.

Also, what would trigger, stop sending emails per ban and start digesting? I'd think some threshold would have to be crossed (bans/minute)?

**Comment by rwaters71** [comment]
This is not a "corner case". I would also love to see some type of aggregated/rate-limited email reports.
My case is the following: Our website forums are getting hit with registration attempts from hundreds of IPs each day, Tor exit relays, etc.  Custom rules in fail2ban are great and allow me to ban them temporarily.  I do want to have email reports of what fail2ban jails are doing and who is getting banned. However, wading through 200-300 emails from fail2ban each morning is quite a nuisance.
In our case, when someone tries to get in the server, we get the same fail2ban rule triggered hundreds of times in the same minute or two, then they are banned an gone for a few hours. However, this generates hundreds of emails in under a minute, it is quite annoying.

IMHO, there should be a setting to buffer/rate-limit similar emails per jail and get ONE digest per user-defined period. This simple setting would save me personally hundreds of unnecessary emails per day. I do want a notification when the rule is triggered, but one with all the IPs, not 200+ emails in a minute. As noted, this is resource-intensive, unnecessary, and does not provide much benefit (vs. a single email per n minutes with all the information). Without rate limiting, I am contemplating turning email notifications off all together, which would prevent me from receiving important fail2ban notifications.

**Comment by sebres** [comment]
@rwaters71 did you read https://github.com/fail2ban/fail2ban/issues/832#issuecomment-60483697?
Would be this sufficient enough?

**Comment by szepeviktor** [comment]
If it is a wishlist I'd like to add a exception mechanism because I treat Hungarian bans differently.

**Comment by rwaters71** [comment]
Yes, it would be more than enough, just not sure we are ready to jump branches ;)

**Comment by rwaters71** [comment]
@szepeviktor  Not a wishlist, I have a similar issue to the original poster.

**Comment by rwaters71** [comment]
@szepviktor I thought that rate-limiting emails would address your original issue as well.

**Comment by leeclemens** [comment]
@rwaters71 I used "corner case" intentionally. How many servers run fail2ban vs. how many fall over sending emails? Before stating "corner case", I also stated "If something like a digest can be implemented, that is great!".

**Comment by leeclemens** [comment]
@szepeviktor What is it about Hungarian bans you treat differently?

**Comment by szepeviktor** [comment]
I live in Hungary and I report/unban/mak phone call about/write email/examine/etc Hungarian bans.

**Comment by rwaters71** [comment]
@leeclemens you are correct of course.. I was trying to clarify that rate-limiting emails would be welcomed by a larger circle of users, sharing my dilemma with the high email reporting rate.

**Comment by leeclemens** [comment]
@szepeviktor I guess I'm just trying to better understand your usecase, particularly as it applies to Hungarian IPs (as you mentioned). Do you unban all Hungarian IPs (i.e. whitelist)? Or would you like to report/email/examine only IPs within only a specific country?

**Comment by szepeviktor** [comment]
> Do you unban all Hungarian IPs (i.e. whitelist)?
Ofcourse not. I investigate all bans: I look up whois data, maybe spam data and examine the application that was abused.
I'd like to *exclude* Hungarian IP-s from digest reporting because I handle them differently.

**Comment by szepeviktor** [comment]
One say the national bank thanked me to report an abuse issue.

**Comment by szepeviktor** [comment]
If anyone interested in low volume notification emails please check out this condition before sendmail:
https://github.com/szepeviktor/debian-server-tools/blob/master/security/fail2ban-leanmail/leanmail.sh

Check IP through the following services and prevent email sending on match.

- Project Honeypot
- Spamhaus XBL
- stopforumspam
- OpenBL
- Cinsscore
- Greensnow
- blocklist.de

There are local IP list files, DNSBL queries and a HTTP API.
IP lists are looked up by matching only the first three octets.

Opinions are welcome.

**Comment by szepeviktor** [comment]
Great success!

```
Oct 17 16:39:31 szerver4 fail2ban-leanmail: Banned IP (136.243.5.215) matches spamhaus
Oct 17 16:39:36 szerver4 fail2ban-leanmail: Banned IP (136.243.75.162) matches spamhaus
Oct 17 16:39:37 szerver4 fail2ban-leanmail: Banned IP (112.198.99.83) matches blde
Oct 17 16:39:38 szerver4 fail2ban-leanmail: Banned IP (211.59.8.170) matches greensnow
Oct 17 16:39:41 szerver4 fail2ban-leanmail: Banned IP (14.37.132.232) matches blde
Oct 17 16:39:46 szerver4 fail2ban-leanmail: Banned IP (162.247.73.206) matches spamhaus
Oct 17 16:39:47 szerver4 fail2ban-leanmail: Banned IP (171.25.193.77) matches greensnow
Oct 17 16:39:48 szerver4 fail2ban-leanmail: Banned IP (176.10.104.240) matches openbl
Oct 17 16:39:48 szerver4 fail2ban-leanmail: Banned IP (176.10.104.243) matches openbl
Oct 17 16:40:19 szerver4 fail2ban-leanmail: Banned IP (176.10.99.207) matches greensnow
Oct 17 16:40:20 szerver4 fail2ban-leanmail: Banned IP (176.31.121.196) matches spamhaus
Oct 17 16:40:21 szerver4 fail2ban-leanmail: Banned IP (185.17.184.228) matches greensnow
Oct 17 16:40:22 szerver4 fail2ban-leanmail: Banned IP (187.114.139.146) matches blde1h
Oct 17 16:40:22 szerver4 fail2ban-leanmail: Banned IP (188.143.234.81) matches greensnow
Oct 17 16:40:23 szerver4 fail2ban-leanmail: Banned IP (192.42.116.16) matches greensnow
Oct 17 16:40:23 szerver4 fail2ban-leanmail: Banned IP (192.99.246.164) matches greensnow
Oct 17 16:40:25 szerver4 fail2ban-leanmail: Banned IP (199.127.226.150) matches blde
Oct 17 16:40:25 szerver4 fail2ban-leanmail: Banned IP (202.21.190.136) matches blde
Oct 17 16:40:26 szerver4 fail2ban-leanmail: Banned IP (212.47.234.38) matches greensnow
Oct 17 16:40:27 szerver4 fail2ban-leanmail: Banned IP (35.0.127.52) matches openbl
Oct 17 16:40:28 szerver4 fail2ban-leanmail: Banned IP (37.187.129.166) matches greensnow
Oct 17 16:40:30 szerver4 fail2ban-leanmail: Banned IP (46.119.125.112) matches greensnow
Oct 17 16:40:30 szerver4 fail2ban-leanmail: Banned IP (5.135.158.101) matches greensnow
Oct 17 16:40:31 szerver4 fail2ban-leanmail: Banned IP (5.135.209.86) matches greensnow
Oct 17 16:40:32 szerver4 fail2ban-leanmail: Banned IP (5.165.67.90) matches spamhaus
Oct 17 16:40:33 szerver4 fail2ban-leanmail: Banned IP (5.196.72.100) matches greensnow
Oct 17 16:40:34 szerver4 fail2ban-leanmail: Banned IP (5.41.253.68) matches blde
Oct 17 16:40:35 szerver4 fail2ban-leanmail: Banned IP (62.210.140.73) matches blde
Oct 17 16:40:37 szerver4 fail2ban-leanmail: Banned IP (82.223.28.73) matches spamhaus
Oct 17 16:40:39 szerver4 fail2ban-leanmail: Banned IP (89.187.142.208) matches greensnow
Oct 17 16:40:42 szerver4 fail2ban-leanmail: Banned IP (93.115.95.206) matches greensnow
Oct 17 16:40:43 szerver4 fail2ban-leanmail: Banned IP (94.53.85.86) matches spamhaus
Oct 17 16:40:44 szerver4 fail2ban-leanmail: Banned IP (95.128.43.164) matches greensnow
Oct 17 16:43:13 szerver4 fail2ban-leanmail: Banned IP (190.80.66.126) matches spamhaus
Oct 17 16:43:34 szerver4 fail2ban-leanmail: Banned IP (31.223.129.142) matches spamhaus
Oct 17 17:01:10 szerver4 fail2ban-leanmail: Banned IP (87.101.253.43) matches blde
```

**Comment by gagomap** [comment]
Why don't you use ipset to block a huge IP list ?

**Comment by szepeviktor** [comment]
@gagomap I am not looking for blocking traffic, that is done by my WordPress WAF + Fail2ban.
It is about the email traffic that Fail2ban generates.

**Comment by maverjk** [comment]
Hello I'm looking for the features  `maillatency = 5m 15m` because I'm keeping having thousands of emails alert when my server got a bruteforce attack on sshd. Any news about it? Thanks

**Comment by szepeviktor** [comment]
Try analyzing your attacking IP-s and find yourself matching lists.
Then you may filter those emails
https://github.com/szepeviktor/debian-server-tools/blob/master/security/fail2ban-leanmail/leanmail.sh

**Comment by maverjk** [comment]
I just would like to use this kind of feature:

>... **It is possible with adding a latency for send a mail, grouping all of bans inside this latency**.
Scenario I have realized, configured maillatency = 5m 15m:
after each ban, it will be added to the report, resets a timer for send email again to 5 minutes;
if no more bans occurred inside a last 5 minutes, email will be send; but at the latest 15 minutes after last sending.
So, it sends a mail with report within interval 5 to 15 minutes;
<

That would be really useful

**Comment by enricotagliavini** [comment]
@maverjk I've implemented something similar to what you describe. It's meant to be a daily digest, but it doesn't have to be daily at all, you can run it every 5 minutes if you want. Possibly just edit the code a bit not to send an email if no event is found.

https://github.com/enricotagliavini/fail2ban-digest

**Comment by szepeviktor** [comment]
Very nice code @enricotagliavini !
Do you plan to add functional and code-styling tests?

**Comment by maverjk** [comment]
@enricotagliavini 
I'm testing your script, really nice job. I have just a couple of points for you that maybe you can consider. 
First in your installation guide you should mention to specify shebang line inside "/bin/fail2ban_digest". Many people not so expert of python script cannot find easy solution at the error "/usr/bin/env: python3: No such file or directory"

Then in "jail.local" I would like to specify "action = %(action_digest)s" as DEFAULT instead to use it for each jail. Is it possible? 
Last consideration, instead to use a cron job to send out digest email, would be really nice to have a timer activated when there is something in db to send out. (ex. as soon as there is something wait 30 minutes to collect all info for each jail).

Of course mine are just consideration and personal opinions. Thanks again to share your solution.

**Comment by enricotagliavini** [comment]
@szepeviktor thank you. I have to say I've been lazy with that. I was simply a bit tired of all the random emails, so I coded that thing together. Functional tests would be a good idea I agree. I used noose in the past and I liked it. For now I have no solid plan, this is just an hobby project for my own private servers. Have to see if I get inspired to work on it more :). It's also very small for now, and sees updates very rarely, so I didn't had much pressure to implement a test suite.

@maverjk thank you. For the shebang there should be no need to change it. Why did you had to change it? I just used the recommended one, see also [1]. With that being said python is installed in many different ways on different systems. On Linux the recommended way is to have /usr/bin/python3, however it's not good to hardcode such a path (e.g. if you want to run from a virtualenv), so you should use env instead, which should exist on any Linux system. Installing python3 and calling it python (without numbers) is not a recommended setup, hence I didn't covered that case.

You should be able to use it as a DEFAULT digest, I never tried it, but I can't imagine a reason why it should not work (and possibly the universe will prove me wrong).

I was thinking about the two approaches, cron vs. timer. In the end I decided for cron for one reason: you know when the email will come, it's not a random time, while with a timer it would still be random, but will aggregate emails for X minutes before sending. I prefer the former approach. This is also because I mainly use it as a daily digest, rather than a 30 minutes digest. I guess for 30 minutes it doesn't matter too much. I could add an option to simply not send an email if there is no event to report. Feel free to open an issue on the project. If I get inspired I might implement it.

[1]: https://docs.python.org/3/using/unix.html?highlight=shebang#miscellaneous
